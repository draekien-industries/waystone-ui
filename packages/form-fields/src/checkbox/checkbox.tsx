'use client';

import {
  type ComponentRef,
  forwardRef,
  useImperativeHandle,
  useRef,
  type ForwardedRef,
  useState,
  useEffect,
} from 'react';
import { Flex } from 'theme-ui';
import { alpha } from '@theme-ui/color';
import type { DisabledAttributes, FullWidthAttributes } from '@waystone/types';
import { Icon } from '@waystone/icon';
import { cssSelectors, outlineCss } from '@waystone/css-presets';
import { Label } from '../label';
import type {
  InputAriaAttributes,
  InputCallbackAttributes,
  SharedInputAttributes,
} from '../types';

export type CheckboxProps = {
  label?: string;
  checked?: boolean;
  defaultChecked?: boolean;
} & InputAriaAttributes &
  InputCallbackAttributes &
  FullWidthAttributes &
  SharedInputAttributes;

export const CheckboxContent = (
  {
    id,
    label,
    name,
    checked,
    defaultChecked,
    required,
    readOnly,
    fullWidth,
    value,
    defaultValue,
    onClick,
    onChange,
    onBlur,
    onFocus,
    ...rest
  }: CheckboxProps,
  ref: ForwardedRef<HTMLInputElement>
) => {
  const inputRef = useRef<ComponentRef<'input'>>(null);
  const buttonRef = useRef<ComponentRef<'button'>>(null);

  const [displayChecked, setDisplayChecked] = useState(defaultChecked);

  useEffect(() => {
    if (checked === undefined) return;

    setDisplayChecked(checked);
  }, [checked]);

  useImperativeHandle<Partial<HTMLInputElement>, Partial<HTMLInputElement>>(
    ref,
    () => ({
      ...inputRef.current,
      focus(options) {
        buttonRef.current?.focus(options);
      },
      blur() {
        buttonRef.current?.blur();
      },
      click() {
        buttonRef.current?.click();
      },
    }),
    []
  );

  return (
    <Flex
      sx={{
        display: fullWidth ? 'flex' : 'inline-flex',
        flexWrap: 'nowrap',
        gap: 'small',
        alignItems: 'center',
        justifyContent: 'flex-start',
      }}>
      <button
        sx={{
          cursor: rest.disabled ? 'not-allowed' : 'pointer',
          width: '1.25rem',
          color: displayChecked ? 'info-400' : 'transparent',
          padding: 0,
          aspectRatio: 1,
          position: 'relative',
          border: '2px solid',
          borderColor: displayChecked ? 'info-400' : 'muted',
          borderRadius: 'small',
          backgroundColor: displayChecked
            ? alpha('info-200', 0.3)
            : alpha('ghost', 0.5),
          ...outlineCss.base,
          [cssSelectors.focusVisible]: {
            ...outlineCss.focused,
          },
          [cssSelectors.hover]: {
            borderColor: displayChecked ? 'info-500' : 'b-400',
          },
          [cssSelectors.disabled]: {
            borderColor: 'ghost',
          },
        }}
        ref={buttonRef}
        type="button"
        role="checkbox"
        aria-checked={displayChecked}
        onClick={() => inputRef.current?.click()}
        {...rest}>
        <Icon size="small" aria-hidden>
          check
        </Icon>
      </button>
      <input
        id={id}
        ref={inputRef}
        type="checkbox"
        hidden
        onClick={(e) => {
          if (onClick) onClick(e);
          setDisplayChecked(!displayChecked);
        }}
        {...{
          name,
          checked,
          defaultChecked,
          required,
          readOnly,
          value,
          defaultValue,
          onFocus,
          onBlur,
          onChange,
        }}
      />
      {label && (
        <Label
          sx={{ cursor: rest.disabled ? 'not-allowed' : 'pointer' }}
          htmlFor={id}>
          {label}
        </Label>
      )}
    </Flex>
  );
};

export const Checkbox = forwardRef(CheckboxContent);
