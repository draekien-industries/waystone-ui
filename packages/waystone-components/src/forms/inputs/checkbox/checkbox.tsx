'use client';

import { HTMLProps, useEffect, useRef } from 'react';

export type CheckboxProps = {
  indeterminate?: boolean;
} & Omit<HTMLProps<HTMLInputElement>, 'type'>;

export const Checkbox = ({
  indeterminate,
  checked,
  ...rest
}: CheckboxProps) => {
  const ref = useRef<HTMLInputElement>(null);

  useEffect(() => {
    if (ref.current) {
      ref.current.indeterminate = !!indeterminate;
    }
  }, [ref, checked, indeterminate]);

  return (
    <input
      sx={{
        cursor: rest.disabled ? 'not-allowed' : 'pointer',
      }}
      type="checkbox"
      ref={ref}
      checked={checked}
      {...rest}
    />
  );
};

export default Checkbox;
