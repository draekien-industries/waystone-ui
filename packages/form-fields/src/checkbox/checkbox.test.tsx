import {
  fireEvent,
  render,
  screen,
  userEvent,
} from '@waystone/testing-library';
import { Checkbox } from './checkbox';

describe('<Checkbox />', () => {
  it('should render a checkbox with no label', () => {
    const { container } = render(
      <Checkbox id="test" data-testid="test-checkbox" />
    );

    expect(container).toMatchSnapshot();
  });

  it('should render a checkbox with a label', () => {
    const { container } = render(
      <Checkbox id="test" data-testid="test-checkbox" label="I am a checkbox" />
    );

    const label = screen.getByText('I am a checkbox');

    expect(label).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should render an input with type checkbox', () => {
    render(<Checkbox id="testing" data-testid="test-checkbox" />);

    const checkbox = screen.getByTestId('test-checkbox');

    expect(checkbox).toHaveAttribute('type', 'checkbox');
  });

  it('should invoke change handler when clicked', async () => {
    const user = userEvent.setup({ delay: null });

    const handleChange = jest.fn();

    render(
      <Checkbox
        id="testing"
        data-testid="test-checkbox"
        onChange={handleChange}
      />
    );

    const checkbox = screen.getByTestId('test-checkbox');

    await user.click(checkbox);

    expect(handleChange).toHaveBeenCalledTimes(1);
  });

  it('should invoke focus handler when focused', () => {
    const handleFocus = jest.fn();

    render(
      <Checkbox
        id="testing"
        data-testid="test-checkbox"
        onFocus={handleFocus}
      />
    );

    const checkbox = screen.getByTestId('test-checkbox');

    fireEvent.focus(checkbox);

    expect(handleFocus).toHaveBeenCalledTimes(1);
  });

  it('should invoke blur handler when blurred', () => {
    const handleBlur = jest.fn();

    render(
      <Checkbox id="testing" data-testid="test-checkbox" onBlur={handleBlur} />
    );

    const checkbox = screen.getByTestId('test-checkbox');

    fireEvent.focus(checkbox);
    fireEvent.blur(checkbox);

    expect(handleBlur).toHaveBeenCalledTimes(1);
  });

  it('should invoke click handler when clicked', () => {
    const handleClick = jest.fn();

    render(
      <Checkbox
        id="testing"
        data-testid="test-checkbox"
        onClick={handleClick}
      />
    );

    const checkbox = screen.getByTestId('test-checkbox');

    fireEvent.click(checkbox);

    expect(handleClick).toHaveBeenCalledTimes(1);
  });
});
