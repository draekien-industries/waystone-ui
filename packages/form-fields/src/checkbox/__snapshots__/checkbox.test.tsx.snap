// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> should render a checkbox with a label 1`] = `
<div>
  <div
    class="css-17w1b6b-CheckboxWithRef"
  >
    <input
      class="css-1pckfla-CheckboxWithRef"
      data-testid="test-checkbox"
      id="test"
      type="checkbox"
    />
    <label
      class="css-vb17rl-CheckboxWithRef"
      for="test"
    >
      I am a checkbox
    </label>
  </div>
</div>
`;

exports[`<Checkbox /> should render a checkbox with no label 1`] = `
<div>
  <div
    class="css-17w1b6b-CheckboxWithRef"
  >
    <input
      class="css-1pckfla-CheckboxWithRef"
      data-testid="test-checkbox"
      id="test"
      type="checkbox"
    />
  </div>
</div>
`;
